### ISSUE TITLE
Blog: `UTF-8` encoding problems on Windows

### ISSUE THREAD

**Issue by avanbremen** [squidfunk/mkdocs-material#6917]
### Bug description

When adding the [built-in blog plugin](https://squidfunk.github.io/mkdocs-material/setup/setting-up-a-blog/#built-in-blog-plugin) and [writing your first post](https://squidfunk.github.io/mkdocs-material/setup/setting-up-a-blog/#writing-your-first-post), spinning up the [live preview server](https://squidfunk.github.io/mkdocs-material/creating-your-site/#previewing-as-you-write) will abort with a `BuildError` (__init__.py [line 73](https://github.com/squidfunk/mkdocs-material/blob/master/src/plugins/blog/structure/__init__.py#L73)):

```
Error reading metadata of post 'blog\posts\hello-world.md' in 'docs':
Expected metadata to be defined but found nothing
```

Including YAML style meta-data in other (not blog) Markdown files will not result in a `BuildError`.

A minimal reproduction project is included.

Tested with [mkdocs-material-9.4.13](https://github.com/squidfunk/mkdocs-material/releases/tag/9.5.13) and [mkdocs-1.5.3](https://github.com/mkdocs/mkdocs/releases/tag/1.5.3).

### Related links

- [Built-in blog plugin](https://squidfunk.github.io/mkdocs-material/setup/setting-up-a-blog/#built-in-blog-plugin)
- [Writing your first post](https://squidfunk.github.io/mkdocs-material/setup/setting-up-a-blog/#writing-your-first-post)

### Reproduction

[material-mkdocs-blog.zip](https://github.com/squidfunk/mkdocs-material/files/14617610/material-mkdocs-blog.zip)

### Steps to reproduce

1. Add [built-in blog plugin](https://squidfunk.github.io/mkdocs-material/setup/setting-up-a-blog/#built-in-blog-plugin)
2. [Write your first post](https://squidfunk.github.io/mkdocs-material/setup/setting-up-a-blog/#writing-your-first-post)
3. Spin up the [live preview server](https://squidfunk.github.io/mkdocs-material/creating-your-site/#previewing-as-you-write)

### Before submitting

- [X] I have read and followed the [bug reporting guidelines](https://squidfunk.github.io/mkdocs-material/contributing/reporting-a-bug/).
- [X] I have attached links to [the documentation](https://squidfunk.github.io/mkdocs-material/), and possibly related [issues](https://github.com/squidfunk/mkdocs-material/issues) and [discussions](https://github.com/squidfunk/mkdocs-material/discussions).
- [X] I assure that I have [removed all customizations](https://squidfunk.github.io/mkdocs-material/contributing/reporting-a-bug/#remove-customizations) before submitting this bug report.
- [x] I have attached a __.zip file__ with a [minimal reproduction](https://squidfunk.github.io/mkdocs-material/guides/creating-a-reproduction/) using the [built-in info plugin](https://squidfunk.github.io/mkdocs-material/plugins/info/).

**Comment by alexvoss** [comment]
I was able to reproduce the error but can also say that this is not generally so. I am just working on a blog tutorial and am writing about adding various things to the header of blog posts, without any problems. The content of the blog post looks fine to me, so I typed out a copy and, hey presto, it worked where the original `hello-world.md` did not.

The file is a UTF-8 file with BOM, which might be causing the problem?

```
$ file hello-world.md
hello-world.md: Unicode text, UTF-8 (with BOM) text, with CRLF line terminators
```

It seems that Python does not read UTF-8 with BOM when the encoding UTF-8 is used? See [answers to this SOF question](https://stackoverflow.com/questions/13590749/reading-unicode-file-data-with-bom-chars-in-python)

**Comment by kamilkrzyskow** [comment]
I haven't run the example, but MkDocs should handle UTF-8-BOM, does the blog plugin replace the source reading logic?

**Comment by alexvoss** [comment]
Unless we declare this an upstream issue and ask the Python community to fix its Unicode support ;o)

**Comment by kamilkrzyskow** [comment]
The current implementation is fixed 😆, back in Python 2 you had to add a `u` before Unicode strings, I doubt they'll make further adjustments. Maybe in Python 4 😏

**Comment by avanbremen** [comment]
@alexvoss @kamilkrzyskow thank you so much for looking into this.

I am using JetBrains Rider on Windows 11. Saving `\docs\blog\posts\hello-world.md` as UTF-8 with NO BOM does indeed fix the issue, good to know 😎. I will probably change Rider settings so that it creates UTF-8 files without the BOM by default.

I noticed that in the barebones project I included the `\docs\index.md` file is also encoded as UTF-8 with BOM. This file does not cause any issue when running `mkdocs serve`. Does this mean this problem is specifically tied to the blog plugin?

Have a great weekend!

**Comment by alexvoss** [comment]
> Does this mean this problem is specifically tied to the blog plugin?

Yes, because the blog plugin reads data from Markdown files itself instead of relying on MkDocs - see the code @kamilkrzyskow pointed to above. 

The same issue could occur on other places. I found other instances where files are read with `utf-8` encoding and not `utf-8-sig`, so this could happen elsewhere. I am sure that @squidfunk will give us his reaction on the matter. 

Will tag this issue as a bug for now. As far as I can see, changing all instances of reading a file from `utf-8` to `utf-8-sig` should fix this throughout the codebase and there should not be any drawbacks. Reading with `utf-8-sig` encoding should read files both with and without BOM. Before we can make those changes we should perhaps do some more due diligence.

**Comment by kamilkrzyskow** [comment]
> I am using JetBrains Rider on Windows 11

Isn't Rider for C#, also weird default, never had an issue with PyCharm, which is also from JetBrain 🤔 

> Does this mean this problem is specifically tied to the blog plugin?

Yes, MkDocs itself does read the files with `utf-8-sig`:
https://github.com/mkdocs/mkdocs/blob/1.5.3/mkdocs/structure/pages.py#L201-L203
https://github.com/mkdocs/mkdocs/blob/master/mkdocs/structure/files.py#L455-L458

the blog plugin's implementation is linked in the previous comment, and based on the comments it was made to provide better support for different data types in the front-matter, as otherwise everything would be a string 🤔

**Comment by squidfunk** [comment]
> Changing the utf-8 to utf-8-sig is the fix in this instance.

I was always asking myself what the `-sig` is for. Now we know 😅 @kamilkrzyskow @alexvoss would one of you like to craft a PR here and in Insiders, so we can replace all instances? Otherwise, I'll do it in the coming days.

**Comment by kamilkrzyskow** [comment]
I should be able to create the PRs today or tomorrow @squidfunk. This only affects file reads, for files, that could be created using another editor. So maybe not every `encoding` needs to be changed to the `-sig` variant. This also doesn't affect only Windows, more of an ancient programming debt when there was the transition from other encodings to UTF-8 as the main standard.

**Comment by squidfunk** [comment]
Perfect, I'll assign you for now. If you run into any troubles, let us know ☺️

**Comment by squidfunk** [comment]
Resolved via #6923 (and Insiders via https://github.com/squidfunk/mkdocs-material-insiders/pull/82)

**Comment by squidfunk** [comment]
Released as part of [9.5.14](https://squidfunk.github.io/mkdocs-material/changelog/#9.5.14).
